pipe_base.py

```diff
@@ -0,0 +1,297 @@
+        any downstream quanta have already been executed (`reset_quanta` can be
+        used to deal with already-executed quanta in advance). Rebuilt quanta
+        and their output datasets are assigned new UUIDs.
+
+        This operation requires read access and temporary-table write access to
+        the central data repository database.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_built_quanta_summary(self) -> Set[tuple[str, DataCoordinate]]:
+        """Report the combinations of task labels and sharding data IDs for
+        which quantum graphs have already been built.
+
+        Empty data IDs are returned for workspaces with no sharding dimensions.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def run_quanta(
+        self,
+        *,
+        quanta: uuid.UUID | Iterable[uuid.UUID] | EllipsisType = ...,
+        where: QuantumGraphExpression | str | EllipsisType = ...,
+        bind: Mapping[str, Any] | None = None,
+        data_id: DataId | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Execute tasks on quanta.
+
+        TODO
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def query_quanta(
+        self,
+        *,
+        where: QuantumGraphExpression | str | EllipsisType = ...,
+        bind: Mapping[str, Any] | None = None,
+        data_id: DataId | None = None,
+        **kwargs: Any,
+    ) -> QuantumGraph:
+        """Query for quanta that have already been built and possibly
+        executed.
+
+        See `Butler.query_provenance` for parameter details.
+
+        Notes
+        -----
+        The returned QuantumGraph is a snapshot of the workspace's state, not a
+        view.  If quanta are currently being executed when this is called, the
+        status for different quanta may not reflect the same instant in time,
+        but the states for a single quantum and its output datasets are always
+        consistent (but possibly already out-of-date by the time the method
+        returns).
+
+        The base class does not specify the result when quanta are currently
+        built while this method is called, but derived classes may enable this.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def accept_failed_quanta(
+        self,
+        *,
+        quanta: uuid.UUID | Iterable[uuid.UUID] | EllipsisType = ...,
+        where: QuantumGraphExpression = ...,
+        bind: Mapping[str, Any] | None = None,
+        data_id: DataId | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Change the status of matching quanta that currently also have status
+        `~QuantumStatus.FAILED` to `~QuantumStatus.SUCCESSFUL`.
+
+        Existing outputs (which should already be marked as
+        `~DatasetStatus.INVALIDATED`) will have their status set to
+        `~DatasetStatus.PRESENT`.
```

Comment:
What is the meaning for dataset to have status `PRESENT` but missing from Datastore?